### PR TITLE
Make log_sum_exp return negative infinity for size 0 inputs

### LINKED ISSUE #1416: Add size_zero checks to log_sum_exp
## Description

The log_sum_exp throws an error if the size of the input is zero.
 
The log_sum_exp in prim and rev needs a `size_zero` check on the input matrix. If the input matrix size is zero handle that case separately. Maybe just return 0 in that case or throw? Not sure what would be the best solution. This is caused by a bug in Eigen's maxCoeff.

This was reported on rstan https://github.com/stan-dev/rstan/issues/689

More investigation and the bugfix was found done on the discourse thread:
https://discourse.mc-stan.org/t/segfault-in-log-sum-exp-function-with-rstan-v2-19-2-for-a-rank-ordered-logit-model/11545

#### Current Version:
v3.0.0

### PR DESCRIPTION
**Pull request by mcol**
## Summary

This fixes #1416 by making log_sum_exp return negative infinity for size 0 inputs.

## Tests

Added test mentioned in https://github.com/stan-dev/math/issues/1416#issuecomment-545579652.

## Side Effects

None.

## Checklist

- [X] Math issue #1416

- [x] Copyright holder: Marco Colombo

    The copyright holder is typically you or your assignee, such as a university or company. By submitting this pull request, the copyright holder is agreeing to the license the submitted work under the following licenses:
      - Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
      - Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

- [X] the basic tests are passing

    - unit tests pass (to run, use: `./runTests.py test/unit`)
    - header checks pass, (`make test-headers`)
    - docs build, (`make doxygen`)
    - code passes the built in [C++ standards](https://github.com/stan-dev/stan/wiki/Code-Quality) checks (`make cpplint`)

- [X] the code is written in idiomatic C++ and changes are documented in the doxygen

- [x] the new changes are tested

### CODE REVIEW

**Comment by mcol** [comment]
This seems to have failed in the `add_vec_mat` portion of `test/unit/math/mix/core/operator_logical_and_test` on Windows. I can't reproduce this failure on Linux. Thoughts?

**Comment by rok-cesnovar** [comment]
This seems to be a similar problem we had last week, some tests keep randomly failing. We need to address this issue separately. 

The only thing we can do in the short term is restarting this stage of the test (I did that just now) and hope the randomness is on our side this time. Sorry for the inconvenience.

File: stan/math/prim/mat/fun/log_sum_exp.hpp [hunk 1]
```diff
@@ -24,6 +25,9 @@ namespace math {
  */
 template <int R, int C>
 double log_sum_exp(const Eigen::Matrix<double, R, C>& x) {
+  if (x.size() == 0)
```
Comment (rok-cesnovar) [inline stan/math/prim/mat/fun/log_sum_exp.hpp]: Small thing, I think its much codign if you add curly brackets in this case

```cpp
if (x.size() == 0) {
   return -std::numeric_limits<double>::infinity();
}
```

File: stan/math/rev/mat/fun/log_sum_exp.hpp [hunk 2]
```diff
@@ -19,6 +19,9 @@ namespace internal {
 
 template <int R, int C>
 inline double log_sum_exp_as_double(const Eigen::Matrix<var, R, C>& x) {
+  if (x.size() == 0)
```
Comment (rok-cesnovar) [inline stan/math/rev/mat/fun/log_sum_exp.hpp]: Same request here.

**Pull request merged**